docs/WorkerAgent.md

```diff
@@ -0,0 +1,44 @@
+Given a Project, we want to be able to run the Lane in a stable, isolated, idempotent manner. We want the user to be able to view the results in real-time and see progress, as well as store the output as an artifact for later access. Ideally this proposal will satisfy the requirements for single instance and distributed architectures for fastlane.ci.
+
+## Proposal
+We create an Agent daemon that is running on each Worker. The Agent can be run by launchd on macOS or systemd on Linux. The Agent should be as small and self-contained as possible with minimum dependencies. It should be very easy to install the Agent on a fresh os install or stock VM image.
+
+The Agent is responsible for the following:
+1. Handle GRPC requests to perform work from Server.
+2. Spawn a process with the appropriate Lane name, configured with the correct execution user, group, and environment.
+3. Connect the output of the command to GRPC response to the Server.
+4. Report back the exit status of the child process.
+5. Deliver artifacts to any external persistence services.
+
+## Process Environment
+Each GRPC request to the Agent should contain all of the necessary configuration information required to run the Lane successfully. This can be accomplished by supplying an environment (the ENV variables) to the child process when it is spawned by the Agent.
+
+<img src="WorkerAgent.svg">
+
+For more on using the environment for configuration (https://12factor.net/config)
+The Agent may have to fetch resources (such as source code) from other services but it should not have to fetch any configuration data.
+
+## Permission model
+The Agent will be required to run arbitrary Ruby code on the Worker. It may also be required to perform operations that require root such as installing Xcode. While there will be other security mechanisms that will enforce strict security against malicious users, we can help ourselves by preventing accidental (or incidental) damage to the Worker by running the Agents under a special user that belongs to the sudoers group. This will allow us to whitelist the commands and permissions the user can perform on a Worker.
+
+## Throttling
+Each agent is capable of spawning as many processes to handle the requests as it wants, but there are stability considerations that we should take into account. Lane execution  may be very resource heavy and launching too many of them may make the Worker kill the process prematurely. There are two possibilities for implementing throttling:
+1. Implement a serial queue on the Agent
+2. The Agent reports its capacity to the Server and rejects requests that cannot be fulfilled.
+
+Given that there will probably be a queueing service in the Server that will be responsible for distributing workload across many workers, it seems like the Agent should not implement another queue. The logic for job queueing should live somewhere higher in the stack and the Agent should simply be able to report its capacity when queried.
```

Comment:
This is great to point out. You're right, the job queueing should live higher up, and the Agent should report its capacity. We should always set this as a capacity of 1, since fastlane can't run in parallel on the same machine. That being said, we might want to define this specifically for capability.
Would it make sense to, instead of saying "capacity for 1" the idea would be "capacity for 1 fastlane job" so that later on we could report back "1 fastlane job, 1 download job, 1 whatever job" so that the Agent could potentially perform other work along with fastlane?
I'd think it could be like
```
 [
  { "fastlane" : 1},
  { "download" : 2},
  { "someOtherJobType" : 1}
]
```

Reply:
I was thinking the same thing, that some jobs maybe be parallelizable and there are a few ways to implement this with different tradeoffs. I didn't feel comfortable committing to any one of them in code at this point, but I did want to document it.

Reply:
🆒